definitions/procedures/foreman/remove_duplicate_obsolete_roles.rb

```diff
@@ -0,0 +1,52 @@
+module Procedures::Foreman
+  class RemoveDuplicateObsoleteRoles < ForemanMaintain::Procedure
+    metadata do
+      for_feature :foreman_database
+      description 'Remove duplicate obsolete roles from DB'
+      confine do
+        check_max_version('foreman', '1.20')
+      end
+    end
+
+    def run
+      duplicate_roles = feature(:foreman_database).query(
+        Checks::Foreman::CheckDuplicateRoles.query_to_get_duplicate_roles
+      )
+      roles_hash = duplicate_roles.each_with_object({}) do |role_rec, new_obj|
+        r_name = role_rec['name']
+        new_obj[r_name] = [] unless new_obj.key?(r_name)
+        new_obj[r_name] << role_rec['id'].to_i
+      end
+      duplicate_role_ids = filter_consumed_roles(roles_hash)
+      remove_obsolete_role_records(duplicate_role_ids)
+    end
+
+    private
+
+    def filter_consumed_roles(roles_hash)
+      consumed_role_ids = find_consumed_role_ids
+      roles_hash.values.map do |ids|
```

Comment:
how about something like this:
```ruby
roles_hash.map do |role_name, ids|
  consumed_ids = ids & consumed_role_ids
  if consumed_ids.count == 1
    ids = consumed_ids
  elsif consumed_ids > 1
    # here we could print message
    # ids -= consumed_ids
    # puts "Assigned duplicates of role #{role_name} can not be automatically fixed."
    # or (preferred)
    ids.delete(consumed_ids.min)
    "update user_roles set role_id = #{consumed_ids.min} where role_id in (#{consumed_ids})"
    # or
    # ids -= consumed_ids
    # update roles set name = concat(name, "-", id) where id in (#{consumed_ids})
  elsif ids.length > 1
    ids.delete(ids.min)
  end
  ids
end.flatten
```
It doesn't add extra complexity and should fix also the rare case.

Reply:
When there is no consumed record for any role present, it's easy to keep a role record with `ids.min`. But when there is more than one consumed records for same role then  user have to decide which one he wants to keep or modify.

Reply:
@kgaikwad I see now, that is probably because of the associated entities, correct? Then we may want to give unique names to the duplicate roles. Something like #name-#id could work to unblock the upgrade and be non-obtrusive. Or print a message such as in the first commented out option in the snippet. What do you think?

Reply:
@mbacovsky, 
+1 
Updated PR with changes using which it will update name of duplicate consumed role records with `name - #id`.